### PR TITLE
added functionality for state_dict/load_state_dict for lr_scheduler ( Fixes: #3026 )

### PR DESCRIPTION
**Pull request by ArmenAg**
I added functionality for majority of lr_schedulers to be able to get and load state_dict. I included tests for these schedulers.

### CODE REVIEW

**Review by fmassa — APPROVED** [review]
This looks generally good to me, thanks!

But before merging, you need to fix the lint issues.

**Comment by ArmenAg** [comment]
@fmassa I fixed the linting issues and removed unused import. Can we merge these now? Thanks!

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -23,6 +22,42 @@ def __init__(self, optimizer, last_epoch=-1):
         self.step(last_epoch + 1)
         self.last_epoch = last_epoch
 
+        self.state = kwargs
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: This will easily get out of sync with attributes if someone changes them after the scheduler is created. It's best to avoid storing redundant information, and implement a specialized `__getstate__` for all schedulers

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -23,6 +22,42 @@ def __init__(self, optimizer, last_epoch=-1):
...
+
+    def __getstate__(self):
+        return {
+            'state': self.state,
+            'last_epoch': self.last_epoch,
+        }
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: IIRC this is the default implementation

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -23,6 +22,42 @@ def __init__(self, optimizer, last_epoch=-1):
...
+        """Loads the schedulers state.
+
+        Arguments:
+            state_dict (dict): scheduler state. Should be an object returned
+                from a call to :meth:`state_dict`.
+        """
+        state = state_dict['state']
+
+        self.last_epoch = state_dict['last_epoch']
+        self.__dict__.update(state)
```
Comment (apaszke) [inline torch/optim/lr_scheduler.py]: I think those two are pretty much equivalent to `__getstate__` and `__setstate__` in this implementation

File: test/test_optim.py [hunk 4]
```diff
@@ -620,6 +620,75 @@ def test_lambda_lr(self):
...
+            scheduler.step()
+        state_dict = scheduler.state_dict()
+
+        scheduler_copy = CosineAnnealingLR(self.opt, T_max=epoch // 2, eta_min=eta_min / 2)
+        scheduler_copy.load_state_dict(state_dict)
+
+        self.assertAlmostEqual(scheduler.last_epoch, scheduler_copy.last_epoch)
+        self.assertAlmostEqual(scheduler.eta_min, scheduler_copy.eta_min)
+        self.assertAlmostEqual(scheduler.T_max, scheduler_copy.T_max)
+        self.assertAlmostEqual(scheduler.get_lr(), scheduler_copy.get_lr())
```
Comment (apaszke) [inline test/test_optim.py]: Can you please isolate a separate helper function for all those tests? I'm thinking of something like
```python
def _check_scheduler_state_dict(self, constr, constr2):
    scheduler = constr()
    for _ in range(epoch):
        scheduler.step()
    scheduler_copy = constr2()
    scheduler_copy.load_state_dict(scheduler.state_dict())
    for key in scheduler.__dict__:
        if key != 'optimizer': continue # don't compare optimizer
        ... # here compare scheduler and the copy on `key`

def test_cosine_lr_state_dict(self):
    self. _check_scheduler_state_dict(
        lambda: CosineAnnealingLR(self.opt, T_max=10, eta_min=1e-10),
        lambda: CosineAnnealingLR(self.opt, T_max=5, eta_min=1e-5)
    )
```

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (ArmenAg) [reply torch/optim/lr_scheduler.py]: Yes. Right. They would be out of sync, but we use the kwargs as a way to monitor what keys the state should contain. Our state_dict function then updates specifically only those keys.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@ArmenAg Once you fix the doc issue, feel free to merge.

**Comment by ArmenAg** [comment]
I fixed the documentation error. Can we merge now?

**Review by apaszke — APPROVED** [review]
